packages/website/private/contents/post-005.mdx

```diff
@@ -0,0 +1,74 @@
+---
+slug: root-element-api
+title: Root element API
+description: Customize your Waku project’s root elements.
+author: sophia
+date: 2024/12/03
+---
+
+As we continue our work on a major rearchitecture for v0.22, we’ve been releasing new features in patch versions. So as the year draws to a close, let’s take a look back over some of these features, starting with the new root element API.
+
+## Root elements
+There are many cases where you may want to customize the attributes of the root `<html>`, `<head>`, or `<body>` elements. You can now do so using the optional root element API. It can be used in both Waku’s file-based “pages router” and the underlying low-level API for programmatic routing.
+
+### File-based router
+The root elements can be rendered via an optional `_root.tsx` file placed in the root of `./src/pages`. It must accept a `children` prop of type `ReactNode`.
+
+```tsx
+// ./src/pages/_root.tsx
+
+// Create root element
+export default async function RootElement({ children }) {
+  return (
+    <html lang="en">
+      <head></head>
+      <body data-version="1.0">{children}</body>
+    </html>
+  );
+}
+
+export const getConfig = async () => {
+  return {
+    render: 'static',
+  };
+};
+```
+
+
+
+### Low-level API
```

Comment:
Can refer to `createPages` as something like "config based routing"? Not necessarily that exactly, but "low-level api" seems more descriptive of something like `defineRouter` to me where the target user is library authors.

Reply:
Yeah, we kind of have two lower level apis. When we say low-level api in README, it's `createPages`. Let's call `defineRouter` minimal api or api for lib authors. I like the idea of calling config-based routing, though.

Reply:
> Let's call defineRouter minimal api or api

I actually confused with defineEntries. Anyways, createPages is a low-level api for app developers and define* is for lib authors.